### PR TITLE
Add/Remove order coupon actions logged in notes

### LINKED ISSUE #30104: 5.4.1: add/remove order coupon should be logged in order notes
> Search tip: You can filter issues using our enhancement label https://github.com/woocommerce/woocommerce/issues?q=is%3Aissue+label%3Aenhancement

No matching issues found https://github.com/woocommerce/woocommerce/issues?q=is%3Aissue+coupon+%22order+notes

**Is your feature request related to a problem? Please describe.**
Some admin in a multi-admin stores incorrectly removes a coupon from a pending payment customer order.
Superviser looking into the matter has no idea if a coupon was ever applied to this order, because coupon add/remove operation leaves no trace of its action anywhere.

**Describe the solution you'd like**
All coupon operations logged in order notes.

**Describe alternatives you've considered**
Store admins manually write an order note for all coupon operations, but we all know that's a process broken by design.

**Additional context**
None.

### PR DESCRIPTION
**Pull request by witlock**
### All Submissions:

* [X] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [X] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Closes #30104 .

### How to test the changes in this Pull Request:

1. Go to WC > Coupons (in the admin) and create a coupon.
2. Go to WC > Orderes (in the admin) and create a manual order, with at least a product.
3. Click "Apply Coupon" to apply the coupon from step 1 to this order.
4. Make sure that the coupon was applied (i.e. amounts update accordingly) and that you see a note in the notes metabox about it.
5. Remove the coupon you just applied by clicking the "x" symbol next to its name.
6. Confirm that amounts are updated accordingly and that a new note appears in the notes metabox indicating that the coupon was removed.

### Other information:

* [X] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [X] Have you successfully run tests with your changes locally?

### FOR PR REVIEWER ONLY:

* [] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: plugins/woocommerce/includes/class-wc-ajax.php [hunk 1]
```diff
@@ -1255,7 +1255,11 @@ public static function remove_order_coupon() {
 				throw new Exception( __( 'Invalid coupon', 'woocommerce' ) );
 			}
 
-			$order->remove_coupon( wc_format_coupon_code( wp_unslash( $coupon ) ) ); // phpcs:ignore WordPress.Security.NonceVerification.Missing, WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
+			$code = wc_format_coupon_code( wp_unslash( $coupon ) ); // phpcs:ignore WordPress.Security.NonceVerification.Missing, WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
+			if ( $order->remove_coupon( $code ) ) {
+				// translators: %s coupon code.
+				$order->add_order_note( sprintf( __( 'Coupon removed: "%s".', 'woocommerce' ), $code ), 0, true );
```
Comment (barryhughes) [inline plugins/woocommerce/includes/class-wc-ajax.php]: I don't think there's a huge risk here, but it might be slightly safer to do:

```suggestion
				$order->add_order_note( esc_html( sprintf( __( 'Coupon removed: "%s".', 'woocommerce' ), $code ), 0, true ) );
```

File: plugins/woocommerce/src/Internal/Orders/CouponsController.php [hunk 2]
```diff
@@ -79,12 +79,16 @@ public function add_coupon_discount( array $post_variables ): object {
...
-		$result = $order->apply_coupon( wc_format_coupon_code( wp_unslash( $coupon ) ) ); // phpcs:ignore WordPress.Security.NonceVerification.Missing, WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
+		$code   = wc_format_coupon_code( wp_unslash( $coupon ) ); // phpcs:ignore WordPress.Security.NonceVerification.Missing, WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
+		$result = $order->apply_coupon( $code );
 
 		if ( is_wp_error( $result ) ) {
 			throw new Exception( html_entity_decode( wp_strip_all_tags( $result->get_error_message() ) ) );
 		}
 
+		// translators: %s coupon code.
+		$order->add_order_note( sprintf( __( 'Coupon applied: "%s".', 'woocommerce' ), $code ), 0, true );
```
Comment (barryhughes) [inline plugins/woocommerce/src/Internal/Orders/CouponsController.php]: Similar to an earlier note:

```suggestion
		$order->add_order_note( esc_html( sprintf( __( 'Coupon applied: "%s".', 'woocommerce' ), $code ), 0, true ) );
```

(Guards against HTML in translations, and potentially against HTML in `$code`—even though we don't expect that in most cases.)

**Review by barryhughes — CHANGES_REQUESTED** [review]
Couple of minor things, otherwise this works per instructions.

**Comment by jorgeatorres** [comment]
Hi @barryhughes 👋

Thanks for the review. I've applied your suggestion about `esc_html()` and also took the opportunity to add auto-refresh of the notes in the sidebar metabox (in line with how the other actions work in this section).

I've updated the testing instructions in the original comment with the details. But in short, you don't have to refresh the page to see the new notes when the coupon is applied/removed.

Please let me know what you think.

File: plugins/woocommerce/client/legacy/js/admin/meta-boxes-order.js [hunk 3]
```diff
@@ -488,6 +488,13 @@ jQuery( function ( $ ) {
 						if ( response.success ) {
 							$( '#woocommerce-order-items' ).find( '.inside' ).empty();
 							$( '#woocommerce-order-items' ).find( '.inside' ).append( response.data.html );
+
+							// Update notes.
+							if ( response.data.notes_html ) {
+								$( 'ul.order_notes' ).empty();
+								$( 'ul.order_notes' ).append( $( response.data.notes_html ).find( 'li' ) );
```
Comment (barryhughes) [inline plugins/woocommerce/client/legacy/js/admin/meta-boxes-order.js:495]: Optional small thing ... I realize the performance benefit is small in real terms:

```suggestion
								$( 'ul.order_notes' )
									.empty()
									.append( $( response.data.notes_html ).find( 'li' ) );
```

Since most methods like these return the same jQuery object back, we can chain them and avoid the cost of re-selecting the same elements. Could also apply to the similar chunk further down.

**Review by barryhughes — APPROVED** [review]
Nice update, works well!

Left a note, however it is an optional change and so I'm hitting approve.

**Comment by jorgeatorres** [comment]
@barryhughes: Thanks for the review 🙇. I've merged this as-is. Hope that's ok.

**Review by lkraav — COMMENTED** [review]
👍💪👏
